### DISCUSSION TITLE
Adaptive update via separate downloader/companion app?

### DISCUSSION THREAD

**Discussion by nils-werner** [rauc/rauc#1440] [Q&A]
Hey,

We're trying to implement an update system for our device, and are considering RAUC for it because it ticks a lot of boxes.

However, as our devices are not connected to any network, we need to run updates via a companion app on a smartphone. We would like the app to download a device update whenever available, and then push the file to the device via Bluetooth. Of course we don't want to always download the entire image, but do an adaptive update whenever possible.

Is there a way to get the adaptive update logic from RAUC into a download tool that's not even on the same device?

**Comment by jluebbe** [comment]
We've thought about related use-cases internally, but haven't had time (or a customer project with the requirement) yet.

Currently, we have this layout when using streaming:
```
+------+     +--------+     +--------+     +------------+     +---------+
|      |     |   DM   |     | Linux  |     |    RAUC    |     | HTTP(S) |
| RAUC | <-> |        | <-> |  NBD   | <-> | NBD / HTTP | <-> |         |
|      |     | Verity |     | Client |     |   Bridge   |     | Server  |
+------+     +--------+     +--------+     +------------+     +---------+
```

To be more flexible, we'd have to support an alternative:

```
+------+     +--------+     +--------+     +----------+     +--------+
|      |     |   DM   |     | Linux  |     |  Custom  |     |        |
| RAUC | <-> |        | <-> |  NBD   | <-> |   NBD    | <-> | Client |
|      |     | Verity |     | Client |     |  Bridge  |     |        |
+------+     +--------+     +--------+     +----------+     +--------+
```

So instead of the built-in NBD/HTTP bridge using curl to fetch ranges from a HTTP server, a separate process would receive requests NBD from RAUC or the kernel and fetch the data using a custom protocol. That could be over WebSocket to some JS running on the user's browser to update from a file with a workflow similar to a DSL router update. Or it could be over Bluetooth to a smartphone app.

The local communication would still be over a UNIX socket. To pass the socket to RAUC, I'd suggest using the [InstallBundle](https://rauc.readthedocs.io/en/latest/reference.html#gdbus-method-de-pengutronix-rauc-installer-installbundle) D-Bus method, by either:

- using a [NBD URI](https://github.com/NetworkBlockDevice/nbd/blob/master/doc/uri.md#nbd-uri-components) pointing to a UNIX socket path in the file system or
- using file descriptor passing via D-Bus in the `args` parameter to InstallBundle (perhaps using just `nbd+unix:///` as the URL to indicate this variant)

In RAUC, the changes should be relatively small, as we just need to use the provided NBD socket instead of using the built-in helper. If you want to implement this yourself, we'd be happy discuss the details via the Matrix/IRC channel and to take a PR. Alternatively, you can also contact us for commercial support via sales@pengutronix.de.

**Reply by nils-werner** [reply]
I'm trying to understand and piece together the bits that you said, and maybe I understood wrong but it sounds like you'd try and tunnel the HTTP streaming request through NBD into the Smartphone client, like a Proxy?

I'm asking because some of our devices are in geographical areas where there is no internet connection, and we were hoping we could pre-download the images, and perform an offline update later. Would this be feasible with this NBD-Bridge scheme?

**Reply by jluebbe** [reply]
> I'm trying to understand and piece together the bits that you said, and maybe I understood wrong but it sounds like you'd try and tunnel the HTTP streaming request through NBD into the Smartphone client, like a Proxy?

Not quite. In RAUC, streaming works by using the NBD client in the kernel to provide a block device with the bundle contents. On top of NBD block dev, we use dm-verity to authenticate each block on-the-fly. So, every access to the bundle (e.g. reading part of a rootfs image) is triggers a NBD read request.

Normally, these NBD read requests are then translated to HTTP range-requests by the unprivileged helper/bridge process implemented in `src/nbd.c`. In your case, the NBD read requests would need to be translated to something else. Perhaps to requests in the custom protocol you use on the bluetooth connection.

> I'm asking because some of our devices are in geographical areas where there is no internet connection, and we were hoping we could pre-download the images, and perform an offline update later. Would this be feasible with this NBD-Bridge scheme?

Yes. Currently, there is no way yet to use a custom bridge instead of the built-in one, but it would be straight-forward to add (i.e. a few days of work).
